[17.0][ADD] delivery_sendcloud_oca

ByteMeAsap opened the pull request:
delivery_sendcloud_oca module provides sendcloud shipping integration with Odoo

This module mostly implements what's described in https://docs.sendcloud.sc/api/v2/shipping/

Full documentation for developers is in https://docs.sendcloud.sc/.

This module works for the Community Edition as well as the Enterprise Edition.

tarteo reviewed (approved):
LGTM! This module was moved and migrated from https://github.com/onesteinbv/addons-sendcloud. We (Onestein) have given permission to transition this to the OCA

rousseldenis commented:
@ByteMeAsap Could you rename this to `delivery_sendcloud_oca`as it exists in Odoo enterprise?

Thanks

rousseldenis commented on delivery_sendcloud_oca/models/onboarding_onboarding.py:
```diff
@@ -0,0 +1,14 @@
+
+
+class Onboarding(models.Model):
+    _inherit = "onboarding.onboarding"
```
@ByteMeAsap @tarteo From where is this inherited?

rousseldenis replied:
Of, found

ByteMeAsap commented:
> @ByteMeAsap Could you rename this to `delivery_sendcloud_oca`as it exists in Odoo enterprise?
> 
> Thanks

Yes, I will do that and create a new PR altogether as for this PR, I had to resolve merge conflicts

ByteMeAsap commented:
> @ByteMeAsap Could you rename this to `delivery_sendcloud_oca`as it exists in Odoo enterprise?
> 
> Thanks

@rousseldenis , Yes , I have renamed the module. Can you help me with how do I go about resolving conflicts as there have been changes in the 17.0 branch since I had raised the PR a while back?

Jan-Onestein reviewed (approved):
Looks good to me

ByteMeAsap committed:
[ADD] delivery_sendcloud_oca

Updated requirements.txt file and changed qty_done to quantity_product_uom for stock move lines

Updates

Updated test scripts

Minor updates

ByteMeAsap commented:
@OCA/delivery-carrier-maintainers, @rousseldenis  , can we have this merged?

Jan-Onestein commented:
Can we please have this merged ??
Can some of you guys push this ??  
( @rousseldenis / @gurneyalex / @jgrandguillaume / @simahawk / @JordiBForgeFlow  / @rafaelbn / @tafaRU / @jbaudoux / @hparfr )

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+
+    @api.depends("sendcloud_service_point_input", "delivery_type")
+    def _compute_sendcloud_service_point_required(self):
+        for carrier in self:
```
`self.update({"sendcloud_service_point_required": False})`

ByteMeAsap replied:
@rousseldenis , I have updated the function

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+    @api.depends()
+    def _compute_sendcloud_country_ids(self):
+        for carrier in self:
+            countries = self.env["sendcloud.shipping.method.country"].search(
```
Do a search before the loop and then filter in it to improve performances.

ByteMeAsap replied:
Done !!

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+
+    @api.onchange("delivery_type")
+    def _onchange_sendcloud_delivery_type(self):
+        self._sendcloud_set_countries()
```
Is this really to be done in an onchange ? This can be bad from a UX point of view.

ByteMeAsap replied:
I have updated to set the countries on write

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+                deleted_parcels.append(parcel_code)
+            elif (
+                res.get("status") == "failed"
+                and res.get("message") == "This shipment is already being cancelled."
```
Could not this be translated ?

ByteMeAsap replied:
No, this cannot be translated

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+            elif res.get("error", {}).get("code") == 404:
+                deleted_parcels.append(parcel_code)  # ignore "Not Found" error
+            elif res.get("error"):
+                raise ValidationError(_("Sendcloud: %s") % res["error"].get("message"))
```
Use translation function arguments instead.

ByteMeAsap replied:
Done

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+    def _compute_can_generate_return(self):
+        res = super()._compute_can_generate_return()
+        for carrier in self.filtered(lambda c: c.delivery_type == "sendcloud"):
+            carrier.can_generate_return = True
```
you can use an update() on recordset instead.

ByteMeAsap replied:
Done !!

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+        ):
+            record.country_ids = record._sendcloud_get_countries_from_cache()
+
+    def _sendcloud_get_countries_from_cache(self):
```
Maybe the method name deos not reflect the current behavior.

ByteMeAsap replied:
Changed the method name

rousseldenis commented on delivery_sendcloud_oca/models/delivery_carrier.py:
```diff
@@ -0,0 +1,476 @@
+            if existing.sendcloud_code not in existing_shipping_methods_map:
+                existing_shipping_methods_map[existing.sendcloud_code] = self.env[
+                    "delivery.carrier"
+                ]
```
`self.env["delivery.carrier"].browse()` is semantically more correct.

rousseldenis commented on delivery_sendcloud_oca/models/sendcloud_shipping_method_country.py:
```diff
@@ -0,0 +1,143 @@
+                shipping_method_country.product_id = item.product_id
+                shipping_method_country.enable_price_custom = item.enable_price_custom
+            else:
+                self.env["sendcloud.shipping.method.country.custom"].create(
```
Build a list of create values in the loop, then call `create()` once.

ByteMeAsap replied:
Done !

rousseldenis reviewed (changes requested):
Code review

ByteMeAsap commented:
@rousseldenis , please review

ByteMeAsap commented:
@pedrobaeza , can we have this merged?

pedrobaeza commented:
@rousseldenis requested changes, so he should now check and remove/approve the review

Jan-Onestein commented:
@rousseldenis ... please review so that it can be merged !! 

@pedrobaeza ... any other way to push it ??  it has been 2 months+

tarteo commented:
@rousseldenis Can you please revise your review?

tarteo commented:
@rousseldenis any update on this?

tarteo commented:
We had to close this, because we can't wait months on end for a response. We will be moving this back to https://github.com/onesteinbv/addons-sendcloud as `delivery_sendcloud_official`

pedrobaeza commented:
Well, let's try a last ping to @rousseldenis

rousseldenis reviewed (approved):
Let's move forward on this

rousseldenis commented:
/ocabot merge nobump
